mmcv/device/ipu/model_converter.py

```diff
@@ -0,0 +1,829 @@
+            self.compare_with_cpu = False
+        else:
+            self.compare_with_cpu = True
+        # move model.fp16_enabled to self.fp16_enabled,
+        # modify the position where the input is automatically casted to half
+        if getattr(model, 'fp16_enabled', False):
+            model.fp16_enabled = False
+            self.fp16_enabled = True
+        # make torch.jit.trace convert self._model
+        model = WrappedNet(model, self.inputs_tree_manager,
+                           self.outputs_tree_manager,
+                           self.hooked_features,
+                           modules_to_record=modules_to_record)
+        super().__init__(model, training=training, *args, **kwargs)
+        # overwrite self._args_parser in train_step or val_step
+        self._args_parser = None
+        if training:
+            assert self.training
+        else:
+            assert not self.training
+
+    @property
+    def training(self):
+        # If trying to get the attribute(training) of self,
+        # since the class has no training attribute,
+        # it will automatically look for the training attribute of self.model.
+        # However, the real attribute we want to check is self._training,
+        # self.model.training  and self._training are often inconsistent.
+        # It is not clear whether it is a Poptorch bug or a special design,
+        # temporarily use this function to fix the problem
+        return self._training  # comes from self.model._training
+
+    @auto_fp16(supported_types=(PoplarExecutor,))
+    def run_model(self, data_dict):
+        # this function is used to parse input_dict
+        # and convert to output_dict
+        if self.isCompiled():
+            self.inputs_tree_manager.set_tree(data_dict)
+            inputs_tuple = tuple(self.inputs_tree_manager.get_tensors())
+        else:
+            # get tensors out of data and put them in a tuple
+            self.inputs_tree_manager.set_tree(data_dict)
+            inputs_tuple = tuple(self.inputs_tree_manager.get_tensors())
+            # turn logger in tree manager off after compilation
+            self.inputs_tree_manager.quick()
+            self.outputs_tree_manager.quick()
+
+        # parser args in the first iter
+        if self._args_parser is None:
+            self._args_parser = DictArgsParser({'args': inputs_tuple})
+
+        # run or convert model
+        # the plain_outputs will be used in converting stage
+        plain_outputs = self(inputs_tuple)
+
+        self.inputs_tree_manager.clean_tensors()
+
+        # put list of tensors back to the output dict
+        # according to the same order
+        self.outputs_tree_manager.set_tensors(plain_outputs)
+        # get the real output dictionary from self.outputs_tree_manager
+        output_dict = self.outputs_tree_manager.tree
+
+        # split output_dict into hooked_features_ipu
+        # and output of the torch model
+        mmcv_model_output = {}
```

Comment:
according to the comments in lines 510&511, it might be better to rename it to torch_model_output